File: 55. Jump Game/src/java/JumpGame.java

public class JumpGame 
{
    public static boolean canJump(int[] nums) throws InterruptedException 
    {
        int index = 0;
        int length = nums.length - 1;

        // for just one element
        if(length == 0)
        {
        	return true;
        }
        
        while(index<length)
        {
        	if(nums[index]>=length)
        	{
        		return true;
        	}
        	//element is 0
        	if(nums[index]==0)
        	{
        		//element is 0, fall back to find
        		int changedindex = index;
        		
        		for(;changedindex>=0;changedindex--)
        		{
        			if(changedindex>0&&nums[changedindex]>0)
        			{
        				//for the element that it can over the left point in 0 position
        				if(nums[changedindex]+changedindex>index)
        				{
        					changedindex += nums[changedindex];
        					index = changedindex;
        					System.out.println("\t"+index);
        					Thread.sleep(1000);
        					break;
        				}
        			}
        		}
        		
        		if(index>=length)
        		{
        			return true;
        		}
        		
        		int loopindex = index-1;
        		
        		//for reverse loop , like 4,3,2,1,0
        		System.out.println("ready");
        		for(;loopindex>=0;loopindex--)
        		{
        			System.out.println("loopindex:"+loopindex);
        			//for reverse loop for index >0
        			if(loopindex>0&&nums[loopindex-1]==nums[loopindex]+1)
        			{
        				//System.out.println("loopindex:"+loopindex);
            			if(nums[loopindex]+loopindex==index&&loopindex==0)
            			{
            				return false;
            			}
        			}
        			//for reverse loop for index == 0
        			else if(loopindex==0&&nums[loopindex+1]==nums[loopindex]-1)
        			{
        				//System.out.println("loopindex:"+loopindex);
            			if(nums[loopindex]+loopindex==index&&loopindex==0)
            			{
            				return false;
            			}
        			}
        		}
        		
        	}
        	index += nums[index];
        	System.out.println(index);
        	Thread.sleep(1000);
        	
    		if(index>=length)
    		{
    			return true;
    		}
    		
        	if(nums[index]==0)
        	{
        		for(int i = 0 ;i <= index;i++)
            	{
            		if(i==index)
            		{
            			return false;
            		}
            		if(nums[i]+i>index)
            		{
            			break;
            		}
            	}	
        	}
        }
        if(index>=length)
        {
        	return true;
        }
        else
        {
        	return false;
        }
    }
}

File: 55. Jump Game/main/JumpGame.py
class Solution:
    def canJump(nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if len(nums) == 1:
            return True
        if nums[0] == 0:
            return False
        left = 0
        right = len(nums) - 1

        while left < right:
            # the index has much bigger number
            if nums[left] > right:
                print(True)
                return True

            # handle the element is 0
            if nums[left] == 0:
                redo_left = left
                # for re-find the bigger number to pass 0 element
                for redo_left in range(left, -1, -1):
                    if nums[redo_left] + redo_left > left:
                        left = nums[redo_left] + redo_left
                        break
                if left >= right:
                    return True

                # handle the element is zero with reverse loop
                reverse_left = left
                for reverse_left in range(left, -1, -1):
                    #print("reverse")
                    if reverse_left > 0 and nums[reverse_left-1] == nums[reverse_left]+1:
                        if reverse_left == 0 and nums[reverse_left] + reverse_left == left:
                            print(False)
                            return False
                    elif reverse_left == 0 and nums[reverse_left+1] == nums[reverse_left] - 1 \
                            and nums[reverse_left] + reverse_left == left:
                        print(False)
                        return False
            elif nums[left] > 0:
                # add the left pointer
                left = nums[left] + left

            if left >= right:
                return True

            #for checking there are enoutgh element to pass the 0
            if nums[left] == 0:
                #print("here")
                i = 0
                for i in range(0, left+1):
                    print(i)
                    print("left"+left.__str__())
                    if i == left:
                        print(False)
                        return False
                    if i + nums[i] > left:
                        break
                    i += 1

